src/function.ts

```diff
@@ -1,43 +1,107 @@
+import {
+  AssetHashType,
+  aws_dynamodb,
+  aws_lambda,
+  CfnResource,
+  DockerImage,
+  Resource,
+  TagManager,
+  Token,
+  Tokenization,
+} from "aws-cdk-lib";
 import * as appsync from "@aws-cdk/aws-appsync-alpha";
-import { aws_lambda } from "aws-cdk-lib";
-import { CallExpr, isVariableReference } from "./expression";
+import { CallExpr, Expr, isVariableReference } from "./expression";
 import { ASL } from "./asl";
+import { Context } from "aws-lambda";
 
 // @ts-ignore - imported for typedoc
 import type { AppsyncResolver, AppSyncVtlIntegration } from "./appsync";
-import { Integration } from "./integration";
+import {
+  NativeFunctionDecl,
+  isNativeFunctionDecl,
+  IntegrationInvocation,
+} from "./declaration";
+import { Construct } from "constructs";
+import { AnyFunction } from "./util";
+import { serializeFunction } from "@functionless/nodejs-closure-serializer";
+import path from "path";
+import fs from "fs";
+import { Err, isErr } from "./error";
+import { IntegrationImpl, Integration } from "./integration";
+import AWS from "aws-sdk";
 
-export function isFunction<P = any, O = any>(a: any): a is Function<P, O> {
+export function isFunction<P = any, O = any>(a: any): a is IFunction<P, O> {
   return a?.kind === "Function";
 }
 
 export type AnyLambda = Function<any, any>;
 
-/**
- * Wraps an {@link aws_lambda.Function} with a type-safe interface that can be
- * called from within an {@link AppsyncResolver}.
- *
- * For example:
- * ```ts
- * const getPerson = new Function<string, Person>(
- *   new aws_lambda.Function(..)
- * );
- *
- * new AppsyncResolver(() => {
- *   return getPerson("value");
- * })
- * ```
- */
-export class Function<P, O> implements Integration {
+export type FunctionClosure<P, O> = (
+  payload: P,
+  context: Context
+) => Promise<O>;
+
+export interface IFunction<P, O> {
+  readonly functionlessKind: typeof Function.FunctionlessType;
+  readonly kind: typeof Function.FunctionlessType;
+  readonly resource: aws_lambda.IFunction;
+
+  (...args: Parameters<ConditionalFunction<P, O>>): ReturnType<
+    ConditionalFunction<P, O>
+  >;
+}
+
+abstract class FunctionBase<P, O>
+  implements IFunction<P, O>, Integration<FunctionBase<P, O>>
+{
   readonly kind = "Function" as const;
+  readonly native: NativeIntegration<FunctionBase<P, O>>;
+  readonly functionlessKind = "Function";
+  public static readonly FunctionlessType = "Function";
 
   readonly appSyncVtl: AppSyncVtlIntegration;
 
   // @ts-ignore - this makes `F` easily available at compile time
   readonly __functionBrand: ConditionalFunction<P, O>;
 
   constructor(readonly resource: aws_lambda.IFunction) {
-    // Integration object for appsync VTL
+    const functionName = this.resource.functionName;
+
+    // Native is used when this function is called from another lambda function's serialized closure.
+    // define this here to closure the functionName without using `this`
```

Comment:
Trying this... maybe we need a lazy?

```ts
let func1: Function<string, string> | undefined;
const func1Lazy = Lazy.integ(() => func1);
const func2 = new Function(async () => {
   func1Lazy.get()("hello")
})
func1 = new Function<string, string>(async () =>{ func2() });
```

This won't work right now because we invoke the integration in the constructor of the Function its used in. 

Or maybe a ref()

```ts
const func1Ref = Function.ref<string, string>();
const funct2 = new Function(async () => {
   func1Ref("");
})
const func1 = func1Ref.set(new Function(async () => { func2() }));
```

Right now... this would just mean calling the bind function and re-writing the environment variables, I think...

https://github.com/functionless/functionless/issues/173